File: homeostasis.js
"use strict";

/**
 * homeostasis.js — Slow Weakening / Pruning Layer (LTD Analog)
 *
 * Biology analog: Long-term depression (LTD) and synaptic pruning.
 * Connections that are never re-activated, never reinforced, and never
 * replayed gradually weaken. Below a floor threshold they are removed
 * entirely, keeping the field sparse and preventing saturation.
 *
 * A cell is eligible for decay if ALL of:
 *   - age > MIN_AGE_DAYS (don't touch freshly stored patterns)
 *   - lastAccessed is stale (not queried recently)
 *   - dreamCount is 0 (never replayed)
 *   - strength < DECAY_STRENGTH_CEILING (only decay weak cells)
 *   - source !== 'promoted-dream' (never decay promoted beliefs)
 *
 * Decay amount scales with how stale and weak the cell already is.
 * Cells that fall below PRUNE_THRESHOLD are removed from the field.
 */

const universe = require('./universe');

const CONFIG = {
    minAgeDays:           1,     // Don't touch cells younger than this
    staleAccessDays:      5,     // "Stale" = not accessed in this many days
    decayStrengthCeiling: 2.0,   // Only decay cells weaker than this
    decayRate:            0.06,  // Strength reduction per homeostasis cycle
    pruneThreshold:       0.09,  // Cells below this are removed entirely
    maxDreamCount:        0,     // Replayed cells are protected
    protectedSources: new Set(['promoted-dream', 'seed']),
};

function _daysSince(ts, now) {
    if (!ts) return Infinity;
    return Math.max(0, (now - ts) / 86400000);
}

function _isDecayEligible(cell, now) {
    // Never decay protected sources
    const src = (cell.meta && cell.meta.source) || '';
    if (CONFIG.protectedSources.has(src)) return false;

    // Must be old enough to consider
    const ageDays = _daysSince(cell.ts, now);
    if (ageDays < CONFIG.minAgeDays) return false;

    // Must be stale (not recently queried)
    const lastAccessDays = _daysSince(cell.lastAccessed || 0, now);
    if (lastAccessDays < CONFIG.staleAccessDays) return false;

    // Must have never been replayed (dream count = 0)
    if ((cell.dreamCount || 0) > CONFIG.maxDreamCount) return false;

    // Only decay genuinely weak cells
    if ((cell.strength || 1) >= CONFIG.decayStrengthCeiling) return false;

    return true;
}

/**
 * runHomeostasis()
 * Scans all cells. Decays eligible ones. Prunes those below floor.
 * Returns { decayed: [...], pruned: [...] }
 */
function runHomeostasis() {
    const cells   = universe.getCells();
    const now     = Date.now();
    const decayed = [];
    const pruned  = [];

    for (const cell of cells) {
        if (!_isDecayEligible(cell, now)) continue;

        const staleFactor = Math.min(1, _daysSince(cell.lastAccessed || 0, now) / 30);
        const weakFactor  = Math.max(0, 1 - (cell.strength / CONFIG.decayStrengthCeiling));
        const amount      = CONFIG.decayRate * (0.5 + staleFactor * 0.3 + weakFactor * 0.2);

        const newStrength = (cell.strength || 1) - amount;

        if (newStrength < CONFIG.pruneThreshold) {
            universe.removeCell(cell.id);
            pruned.push({ id: cell.id, text: cell.text, region: cell.region, finalStrength: cell.strength });
        } else {
            universe.reinforceCell(cell.id, -amount, { lastDecayed: now });
            decayed.push({ id: cell.id, text: cell.text, newStrength });
        }
    }

    return { decayed, pruned };
}

module.exports = {
    CONFIG,
    runHomeostasis,
};

File: kai-rust/src/cognition/homeostasis.rs
/// Homeostasis — Slow weakening / pruning (LTD analog).
///
/// Connections that are never re-activated, reinforced, or replayed
/// gradually weaken. Below a floor threshold they are removed,
/// keeping the field sparse and preventing saturation.
use crate::core::Universe;

pub struct HomeostasisConfig {
    pub min_age_secs: u64,
    pub stale_access_secs: u64,
    pub decay_strength_ceiling: f32,
    pub decay_rate: f32,
    pub prune_threshold: f32,
}

impl Default for HomeostasisConfig {
    fn default() -> Self {
        Self {
            min_age_secs: 86400,          // 1 day
            stale_access_secs: 5 * 86400, // 5 days
            decay_strength_ceiling: 2.0,
            decay_rate: 0.06,
            prune_threshold: 0.09,
        }
    }
}

#[derive(Debug)]
pub struct HomeostasisResult {
    pub decayed: usize,
    pub pruned: usize,
}

/// Run homeostasis: decay stale weak cells, prune below threshold.
pub fn run_homeostasis(universe: &mut Universe, config: &HomeostasisConfig) -> HomeostasisResult {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();

    let mut decayed = 0;
    let mut to_remove = Vec::new();

    for (i, cell) in universe.cells().iter().enumerate() {
        // Never decay promoted beliefs or seeds
        if cell.source == "promoted-dream" || cell.source == "seed" {
            continue;
        }

        // Must be old enough
        let age = now.saturating_sub(cell.created);
        if age < config.min_age_secs {
            continue;
        }

        // Only decay weak cells
        if cell.strength >= config.decay_strength_ceiling {
            continue;
        }

        // Compute decay amount
        let stale_factor = (age as f32 / (30.0 * 86400.0)).min(1.0);
        let weak_factor = (1.0 - cell.strength / config.decay_strength_ceiling).max(0.0);
        let amount = config.decay_rate * (0.5 + stale_factor * 0.3 + weak_factor * 0.2);

        let new_strength = cell.strength - amount;
        if new_strength < config.prune_threshold {
            to_remove.push(i);
        } else {
            decayed += 1;
        }
    }

    // Apply decay to cells not being removed
    let cells = universe.cells_mut();
    for (i, cell) in cells.iter_mut().enumerate() {
        if to_remove.contains(&i) {
            continue;
        }
        if cell.source == "promoted-dream" || cell.source == "seed" {
            continue;
        }

        let age = now.saturating_sub(cell.created);
        if age < config.min_age_secs {
            continue;
        }
        if cell.strength >= config.decay_strength_ceiling {
            continue;
        }

        let stale_factor = (age as f32 / (30.0 * 86400.0)).min(1.0);
        let weak_factor = (1.0 - cell.strength / config.decay_strength_ceiling).max(0.0);
        let amount = config.decay_rate * (0.5 + stale_factor * 0.3 + weak_factor * 0.2);
        cell.strength = (cell.strength - amount).max(0.0);
    }

    // Remove pruned cells (reverse order to maintain indices)
    let pruned = to_remove.len();
    for &i in to_remove.iter().rev() {
        cells.remove(i);
    }

    HomeostasisResult { decayed, pruned }
}
